File: src/main/java/GildedRose.java
import java.util.ArrayList;
import java.util.List;


public class GildedRose {

    private static List<Item> items = null;

    /**
     * @param args
     */
    public static void main(String[] args) {

        System.out.println("OMGHAI!");

        items = new ArrayList<Item>();
        items.add(new Item("+5 Dexterity Vest", 10, 20));
        items.add(new Item("Aged Brie", 2, 0));
        items.add(new Item("Elixir of the Mongoose", 5, 7));
        items.add(new Item("Sulfuras, Hand of Ragnaros", 0, 80));
        items.add(new Item("Backstage passes to a TAFKAL80ETC concert", 15, 20));
        items.add(new Item("Conjured Mana Cake", 3, 6));

        updateQuality();
    }


    public static void updateQuality() {
        for (Item item : items) {
            if ("Sulfuras, Hand of Ragnaros".equals(item.getName())) {
                continue;
            }
            if (("Aged Brie".equals(item.getName()))) {
                increaseQuality(item);
            } else if ("Backstage passes to a TAFKAL80ETC concert".equals(item.getName())) {
                increaseQuality(item);

                if (item.getSellIn() < 11) {
                    increaseQuality(item);
                }

                if (item.getSellIn() < 6) {
                    increaseQuality(item);
                }
            } else {
                if (item.getQuality() > 0) {
                    item.setQuality(item.getQuality() - 1);
                    System.out.println("Quality -1: " + dump(item));
                }
            }

            item.setSellIn(item.getSellIn() - 1);
            System.out.println("SellIn -1: " + dump(item));

            if (item.getSellIn() < 0) {
                if ("Aged Brie".equals(item.getName())) {
                    increaseQuality(item);
                } else if ("Backstage passes to a TAFKAL80ETC concert".equals(item.getName())) {
                    item.setQuality(0);
                    System.out.println("Quality -> 0: " + dump(item));
                } else {
                    if (item.getQuality() > 0) {
                        item.setQuality(item.getQuality() - 1);
                        System.out.println("Quality -1: " + dump(item));
                    }
                }
            }
        }

    }

    private static void increaseQualityBy() {

    }

    private static void increaseQuality(final Item item) {
        if (item.getQuality() < 50) {
            item.setQuality(item.getQuality() + 1);
            System.out.println("Quality +1: " + dump(item));
        }
    }

    private static String dump(Item item) {
        return String.format("[name:%s;sellIn:%s;quality:%s]", item.name, item.sellIn, item.quality);
    }
}

File: gilded-rose/scala/src/main/scala/de/britter/gildedrose/GildedRose.scala
/*
 * Copyright 2016 Benedikt Ritter
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package de.britter.gildedrose

class GildedRose(val items: Item*) {

  def updateQuality() {
    for (item <- items) {
      case Item("Sulfuras, Hand of Ragnaros", _, _) =>
      case item@Item("Aged Brie", sellIn, quality) if quality < 50 && sellIn < 0 => 
        item.sellIn -= 1
        item.quality += 2
      case item@Item("Aged Brie", sellIn, quality) if quality < 50 && sellIn >= 0 => 
        item.sellIn -= 1
        item.quality += 1
      case item@Item("Backstage passes to a TAFKAL80ETC concert", sellIn, quality) if quality < 50 && sellIn > 10 =>
        item.sellIn -= 1
        item.quality += 1
      case item@Item("Backstage passes to a TAFKAL80ETC concert", sellIn, quality) if quality < 50 && sellIn <= 10 && item.sellIn > 5 =>
        item.sellIn -= 1
        item.quality += 2
      case item@Item("Backstage passes to a TAFKAL80ETC concert", sellIn, quality) if quality < 50 && sellIn <= 5 && item.sellIn > 0 =>
        item.sellIn -= 1
        item.quality += 3
      case item@Item("Backstage passes to a TAFKAL80ETC concert", _, _) =>
        item.sellIn -= 1
        item.quality = 0
      case item@Item(name, sellIn, quality) if name.startsWith("Conjured") && quality > 0 && sellIn > 0 =>
        item.sellIn -= 1
        item.quality -= 2
      case item@Item(name, sellIn, quality) if name.startsWith("Conjured") && quality > 0 =>
        item.sellIn -= 1
        item.quality -= 4
      case item@Item(_, sellIn, quality) if quality > 0 && sellIn > 0 =>
        item.sellIn -= 1
        item.quality -= 1
      case item@Item(_, sellIn, quality) if quality > 0 =>
        item.sellIn -= 1
        item.quality -= 2
    }
  }

}
